text/0000-multiarch-builders-and-package.md

```diff
@@ -0,0 +1,944 @@
+
+For those buildpack authors that are using `cross-compile` languages like [go](https://go.dev/) or [rust](https://www.rust-lang.org/) or maybe bash scripts, adding the capability to `pack buildpack package` and `pack builder create` to create multi-arch images 
+and also handle the creation of an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md) will simplify their CI/CD pipelines and make the experience more suitable.
+
+- What is the expected outcome?
+
+`pack buildpack package` and `pack builder create` commands will be updated in a way that they will handle the creation of multi-arch OCI images
+
+# What it is
+[what-it-is]: #what-it-is
+
+The end-users for this proposal are **Buildpack authors**, we expect to improve their user experience when creating multi-architecture buildpacks and builders as follows
+
+## Multi-arch example
+
+Let's suppose a **Buildpack author** has a `buildpack.toml` created with a content similar to this one:
+
+```toml
+# Buildpack API version
+api = "0.12"
+
+# Buildpack ID and metadata
+[buildpack]
+  id = "examples/my-multiarch-buildpack"
+  version = "0.0.1"
+
+# List of targets operating systems, architectures and versions
+
+[[targets]]
+os = "linux"
+arch = "amd64"
+
+[[targets]]
+os = "linux"
+arch = "arm64"
+
+[[targets]]
+os = "windows"
+arch = "amd64"
+
+[[targets.distributions]]
+name = "windows"
+versions = ["10.0.20348.1970"]
+```
+
+And organizes the binaries according to their os/arch with a structure similar to this one:
+
+```bash
+my-multiarch-buildpack
+├── buildpack.toml
+├── linux-amd64
+│         └── bin
+│             ├── build
+│             └── detect
+├── linux-arm64
+│         └── bin
+│             ├── build
+│             └── detect
+└── windows-amd64
+    └── 10.0.20348.1970
+        └── bin
+            ├── build.bat
+            └── detect.bat
+```
+
+Now `pack` will be able to package them separately for each os/arch family, following our [guide](https://buildpacks.io/docs/buildpack-author-guide/package-a-buildpack/) 
+we will need to create a `package.toml` file.
+
+The `package.toml` file will be similar to the one describe in our documentation:
+
+```toml
+[buildpack]
+uri = "examples/my-multiarch-buildpack"
+# OR a .tgz with the previous folder structure
+uri = "my-multiarch-buildpack.tgz"
+```
+
+But in this case, we actually **remove** the [platform](https://buildpacks.io/docs/reference/config/package-config/) section because it will be taken from the `buildpack.toml`.
+
+Packaging a multi-arch buildpack will require the output to be **publish** to a registry or **saved on disk** in OCI layout format.
+
+```bash
+pack buildpack package my-buildpack --config ./package.toml --publish --multi-arch
+# Or
+pack buildpack package my-buildpack.cnb --config ./package.toml --format file --multi-arch
```

Comment:
Could the `--multi-arch` flag be inferred? For example, if `buildpack.toml` contains multiple target/arch's then this flag is automatic. Or are we worried someone might have that but still only want to build a buildpack for a single arch?

Reply:
It is exactly this: 
> Or are we worried someone might have that but still only want to build a buildpack for a single arch?
At first, I was thinking on doing it without the `--multi-arch` flag, but then I decided to force the user to give the acknowledgment

Reply:
I agree that having to type `--multi-arch` everywhere is cumbersome. I think if the buildpack author took the trouble to arrange their binaries according to the nested directory structure then we should just infer that they want a multi-arch image. Or maybe there could be a `pack config` option that sets multi-arch (or not) as the default.

Reply:
I removed the `--multi-arch` flag